pol-core/ecompile/EfswFileWatchListener.cpp

```diff
@@ -0,0 +1,175 @@
+// `watched_files`.
+bool EfswFileWatchListener::add_file( const fs::path& filepath )
+{
+  return watched_files.emplace( filepath ).second;
+}
+
+
+// Adds a directory to watched directories. Returns `true` if the directory was newly added to
+// `watched_dirs`.
+bool EfswFileWatchListener::add_dir( const std::filesystem::path& dir )
+{
+  return watched_dirs.emplace( dir ).second;
+}
+
+
+// Adds a directory to the filesystem watcher. Returns `true` if the directory
+// has a newly added watchID added to `dir_to_watchid`.
+//
+// Throws if the filesystem watcher fails to add the watch.
+bool EfswFileWatchListener::add_watch_dir( const std::filesystem::path& dir )
+{
+  if ( auto itr = dir_to_watchid.find( dir ); itr == dir_to_watchid.end() )
+  {
+    auto watchID = watcher.addWatch( dir.generic_string(), this, true );
+    if ( watchID < 0 )
+    {
+      throw std::runtime_error( "Failed to watch " + dir.generic_string() );
+    }
+    dir_to_watchid.emplace( dir, watchID );
+    return true;
+  }
+
+  return false;
+}
+
+// Removes a file from watched files. Returns `true` if the file was removed
+// from `watched_files`.
+bool EfswFileWatchListener::remove_file( const std::filesystem::path& filepath )
+{
+  return watched_files.erase( filepath ) == 1;
+}
+
+// Returns `true` if a file is watched. A file is watched if it is explicitly inside
+// `watched_files`, or it is in a subdirectory of any directory inside
+// `watched_dirs`.
+bool EfswFileWatchListener::is_watched( const std::filesystem::path& filepath )
+{
+  if ( watched_files.find( filepath ) != watched_files.end() )
+    return true;
+
+  auto dirpath = filepath.parent_path();
+  auto ext = fs::path( filepath ).extension();
+
+  if ( extension_filter.find( ext ) != extension_filter.end() )
+  {
+    auto dirname = dirpath.generic_string();
+    for ( const auto& dir : watched_dirs )
+    {
+      if ( dirname.rfind( dir.generic_string(), 0 ) == 0 )
+      {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+// Adds a `message` to the message queue. The messages is only added if there is
+// not one of the same `filepath` and `old_filepath`.
+void EfswFileWatchListener::add_message( WatchFileMessage&& message )
+{
+  bool add = true;
+  {
+    std::lock_guard<std::mutex> lock( mutex );
+    for ( const auto& other : messages )
+    {
+      {
+        if ( other.filepath == message.filepath && other.old_filepath == message.old_filepath )
+        {
+          add = false;
+          break;
+        }
+      }
+    }
+    if ( add )
+    {
+      messages.push_back( message );
+      handle_messages_by = Clib::wallclock() + DEBOUNCE;
+    }
+  }
+}
+
+// Removes the messages and inserts them into the provided `to_message` list.
+// The messages are only removed if the `handle_messages_by` time clock has
+// passed.
+void EfswFileWatchListener::take_messages( std::list<WatchFileMessage>& to_messages )
+{
+  std::lock_guard<std::mutex> lock( mutex );
```

Comment:
conditional_variable would remove the current "busy" loop with the sleep of 1s

Reply:
We still need the listener for Ctrl+C which is why there's the 1s wait. Can that be achieved with cond vars?

Reply:
Found `cv.wait_for`, :+1:

Reply:
Resolved in https://github.com/polserver/polserver/pull/630/commits/c5a2fa7f7b553a333c43fb5f1d48526bb2f88807